gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java

```diff
@@ -0,0 +1,201 @@
+      LOGGER.warn("No JobSpec with URI:" + uri + " is found.");
+      return null;
+    } else {
+      return resultJobSpecList.get(0);
+    }
+  }
+
+  /**
+   * For each new coming JobCatalogListener, react accordingly to add all existing JobSpec.
+   * @param jobListener
+   */
+  @Override
+  public synchronized void addListener(JobCatalogListener jobListener) {
+    Preconditions.checkNotNull(jobListener);
+
+    this.listeners.addListener(jobListener);
+
+    List<JobSpec> currentJobSpecList = this.getJobs();
+    if (currentJobSpecList == null || currentJobSpecList.size() == 0) {
+      return;
+    } else {
+      for (JobSpec jobSpecEntry : currentJobSpecList) {
+        AddJobCallback addJobCallback = new AddJobCallback(jobSpecEntry);
+        this.listeners.callbackOneListener(addJobCallback, jobListener);
+      }
+    }
+  }
+
+  @Override
+  public synchronized void removeListener(JobCatalogListener jobListener) {
+    this.listeners.removeListener(jobListener);
+  }
+
+  /**
+   * return the path that it controls.
+   * @return
+   */
+
+  public Path getPath() {
+    return this.jobConfDirPath;
+  }
+
+  /**
+   * Monitor specially for file system as it is stateful storage system, which might result in monitoring loop.
+   * Note that here the entity for monitoring is job conf. file instead of JobSpec Object.
+   */
+  private void startGenericFSJobConfigMonitor()
+      throws Exception {
+    FSPathAlterationListenerAdaptor configFilelistener = new FSPathAlterationListenerAdaptor(this.jobConfDirPath);
+    JobCatalogUtils.addPathAlterationObserver(this.pathAlterationMonitor, configFilelistener, this.jobConfDirPath);
+    this.pathAlterationMonitor.start();
+  }
+
+  private class FSPathAlterationListenerAdaptor extends PathAlterationListenerAdaptor {
+    Path jobConfigFileDirPath;
+
+    FSPathAlterationListenerAdaptor(Path jobConfigFileDirPath) {
+      this.jobConfigFileDirPath = jobConfigFileDirPath;
+    }
+
+    /**
+     * Transform the event triggered by file creation into JobSpec Creation for Driver (One of the JobCatalogListener )
+     * Create a new JobSpec object and notify each of member inside JobCatalogListenersList
+     * @param path
+     */
+    @Override
+    public void onFileCreate(Path path) {
+      JobSpec newJobSpec = JobCatalogUtils.loadJobConfigHelper(jobConfDirPath, Action.SINGLEJOB, path).get(0);
+      AddJobCallback addJobCallback = new AddJobCallback(newJobSpec);
+      listeners.callbackAllListeners(addJobCallback);
+    }
+
+    /**
+     * In the call to {@link UpdateJobCallback}, it is hard to retrieve oldJobSpec without caching layer suppport.
+     * Here simply passed the null.
+     * @param path
+     */
+    @Override
+    public void onFileChange(Path path) {
+      JobSpec updatedJobSpec = JobCatalogUtils.loadJobConfigHelper(jobConfDirPath, Action.SINGLEJOB, path).get(0);
+      UpdateJobCallback updateJobCallback = new UpdateJobCallback(null, updatedJobSpec);
```

Comment:
The original job shouldn't be `null`, but on the other hand I wouldn't want the catalog to have to remember  all jobs... let's talk offline.